core/consensus/rhd/src/lib.rs

```diff
@@ -670,20 +914,438 @@ pub fn sign_message<B: Block + Clone>(
+
+		let mut next_index = {
+			let local_id = self.local_key.public().0;
+			let cur_index = self.transaction_pool.cull_and_get_pending(&BlockId::hash(self.parent_hash), |pending| pending
+				.filter(|tx| tx.verified.sender == local_id)
+				.last()
+				.map(|tx| Ok(tx.verified.index()))
+				.unwrap_or_else(|| self.client.account_nonce(&self.parent_id, local_id))
+				.map_err(Error::from)
+			);
+
+			match cur_index {
+				Ok(cur_index) => cur_index + 1,
+				Err(e) => {
+					warn!(target: "consensus", "Error computing next transaction index: {:?}", e);
+					return;
+				}
+			}
+		};
+
+		for (target, misbehavior) in misbehavior {
+			let report = MisbehaviorReport {
+				parent_hash: self.parent_hash.into(),
+				parent_number: self.parent_number.as_(),
+				target,
+				misbehavior: match misbehavior {
+					GenericMisbehavior::ProposeOutOfTurn(_, _, _) => continue,
+					GenericMisbehavior::DoublePropose(_, _, _) => continue,
+					GenericMisbehavior::DoublePrepare(round, (h1, s1), (h2, s2))
+						=> MisbehaviorKind::BftDoublePrepare(round as u32, (h1.into(), s1.signature), (h2.into(), s2.signature)),
+					GenericMisbehavior::DoubleCommit(round, (h1, s1), (h2, s2))
+						=> MisbehaviorKind::BftDoubleCommit(round as u32, (h1.into(), s1.signature), (h2.into(), s2.signature)),
+				}
+			};
+			let payload = (next_index, Call::Consensus(ConsensusCall::report_misbehavior(report)), Era::immortal(), self.client.genesis_hash());
+			let signature = self.local_key.sign(&payload.encode()).into();
+			next_index += 1;
+
+			let local_id = self.local_key.public().0.into();
+			let extrinsic = UncheckedExtrinsic {
+				signature: Some((node_runtime::RawAddress::Id(local_id), signature, payload.0, Era::immortal())),
+				function: payload.1,
+			};
+			let uxt: <<C as AuthoringApi>::Block as BlockT>::Extrinsic = Decode::decode(&mut extrinsic.encode().as_slice()).expect("Encoded extrinsic is valid");
+			let hash = BlockId::<<C as AuthoringApi>::Block>::hash(self.parent_hash);
+			if let Err(e) = self.transaction_pool.submit_one(&hash, uxt) {
+				warn!("Error importing misbehavior report: {:?}", e);
+			}
+		}
+	}
+
+	fn on_round_end(&self, round_number: u32, was_proposed: bool) {
+		let primary_validator = self.validators[
+			self.primary_index(round_number, self.validators.len())
+		];
+
+		// alter the message based on whether we think the empty proposer was forced to skip the round.
+		// this is determined by checking if our local validator would have been forced to skip the round.
+		if !was_proposed {
+			let public = ed25519::Public::from_raw(primary_validator.0);
+			info!(
+				"Potential Offline Validator: {} failed to propose during assigned slot: {}",
+				public,
+				round_number,
+			);
+		}
+
+		self.offline.write().note_round_end(primary_validator, was_proposed);
+	}
+}
+
+fn current_timestamp() -> u64 {
+	time::SystemTime::now().duration_since(time::UNIX_EPOCH)
+		.expect("now always later than unix epoch; qed")
```

Comment:
Why do we claim here that `now` is always later than UNIX epoch but in aura code we try to handle it?